### PR TITLE
add W&B sweeps

### PR DESCRIPTION
**Pull request by AyushExel**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1742,6 +1750,8 @@ def hyperparameter_search(
             )
         if backend == HPSearchBackend.SIGOPT and not is_sigopt_available():
             raise RuntimeError("You picked the sigopt backend, but it is not installed. Use `pip install sigopt`.")
+        if backend == HPSearchBackend.WANDB and not is_wandb_available():
+            raise RuntimeError("You picked the sigopt backend, but it is not installed. Use `pip install sigopt`.")
```
Comment (morganmcg1) [inline src/transformers/trainer.py]: Just need to update this to reference wandb instead of sigopt

**Comment by AyushExel** [comment]
@borisdayma @morganmcg1 can you please review this?

**Comment by borisdayma** [comment]
Do you have example runs: a run in colab and a run as a script?

**Comment by AyushExel** [comment]
@borisdayma here's an example [script ](https://wandb.ai/cayush/good_sweeepXDXOXOXOXD/runs/61dzzqy7/code?workspace=user-cayush). This should also work in a colab

File: src/transformers/integrations.py [hunk 2]
```diff
@@ -124,6 +124,10 @@ def hp_params(trial):
     if is_sigopt_available():
         if isinstance(trial, dict):
             return trial
+    
+    if is_wandb_available():
+        if isinstance(trial, dict):
```
Comment (borisdayma) [inline src/transformers/integrations.py:129]: Should we also update `default_hp_search_backend`?

File: src/transformers/integrations.py [hunk 3]
```diff
@@ -337,6 +341,36 @@ def run_hp_search_sigopt(trainer, n_trials: int, direction: str, **kwargs) -> Be
...
+
+    sweep_config = trainer.hp_space(None)
+    sweep_config['metric']['goal'] = direction
+    sweep_config['name'] = name
+
+    sweep_id = wandb.sweep(sweep_config, project=project, entity=entity) if not sweep_id else sweep_id
+    wandb.agent(sweep_id, function=_objective, count=n_trials)
+
+    #best_trial = study.best_trial
+    #return BestRun(str(best_trial.number), best_trial.value, best_trial.params)
```
Comment (borisdayma) [inline src/transformers/integrations.py]: We probably need to return a best run as well to be compatible with current API

File: src/transformers/integrations.py [hunk 4]
```diff
@@ -337,6 +341,36 @@ def run_hp_search_sigopt(trainer, n_trials: int, direction: str, **kwargs) -> Be
     return best_run
 
 
+def run_hp_search_wandb(trainer, n_trials: int, direction: str, **kwargs) -> BestRun:
+    import wandb
+
+    def _objective():
+        run = wandb.init()
```
Comment (borisdayma) [inline src/transformers/integrations.py]: We need to see if it creates any issue with the wandb callback, both in script and notebooks (in particular closing a run that should not be closed).

Ideally we won't start a run there and have it handled by the `setup` method. We can check and confirm that a run exists or force the report to wandb.

**Comment by AyushExel** [comment]
This is a working [colab](https://colab.research.google.com/drive/1ntpJChI-adiQqhxs-UJQrKwPP6IXpdI9?usp=sharing). And the [sweep ](https://wandb.ai/cayush/good_sweeep_colab/sweeps/4mvwe55a?workspace=user-cayush)from this colab

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -913,6 +918,12 @@ def _report_to_hp_search(
             if self.control.should_save:
                 self._tune_save_checkpoint()
             tune.report(objective=self.objective, **metrics)
+        elif self.hp_search_backend == HPSearchBackend.WANDB:
+            import wandb
+
+            metrics['objective'] = self.objective
```
Comment (borisdayma) [inline src/transformers/trainer.py]: Is it needed or will the objective already be part of metrics (for example `eval_accuracy` or `eval_loss`)?

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -1238,7 +1249,7 @@ def train(
         self.callback_handler.lr_scheduler = self.lr_scheduler
         self.callback_handler.train_dataloader = train_dataloader
         self.state.trial_name = self.hp_name(trial) if self.hp_name is not None else None
-        self.state.trial_params = hp_params(trial.assignments) if trial is not None else None
+        #self.state.trial_params = hp_params(trial.assignments) if trial is not None else None
```
Comment (borisdayma) [inline src/transformers/trainer.py]: I think this is mainly to save the Trainer State and have a record of the parameters.

**Review by borisdayma — COMMENTED** [review]
Looks great!

File: src/transformers/trainer.py (hunk 5 above)
Comment (AyushExel) [reply src/transformers/trainer.py]: Yeah, this is something I'm not sure about. What is the user provides custom objective function? It'll have no name, so we won't be able to tell sweeps API to track that metric. See out DM with Morgan for more discussion on this

File: src/transformers/integrations.py (hunk 2 above)
Comment (AyushExel) [reply src/transformers/integrations.py:129]: Yup. Will do

File: src/transformers/integrations.py (hunk 3 above)
Comment (AyushExel) [reply src/transformers/integrations.py]: Can we get best run from the sweeps agent?

File: src/transformers/integrations.py (hunk 4 above)
Comment (AyushExel) [reply src/transformers/integrations.py]: Yup. Testing this on a colab now

File: src/transformers/trainer.py (hunk 6 above)
Comment (AyushExel) [reply src/transformers/trainer.py]: Yeah, but this seems incompatible with wandb sweep parameter dictionary. It throws KeyError. Also, this variable isn't used anywhere in throughout the module

File: src/transformers/integrations.py (hunk 3 above)
Comment (borisdayma) [reply src/transformers/integrations.py]: We can either:
* [export run summary](https://docs.wandb.ai/guides/track/public-api-guide#querying-multiple-runs) and filter by sweep id
* keep track of the local best run (so result may be different per agent)

File: src/transformers/trainer.py (hunk 6 above)
Comment (borisdayma) [reply src/transformers/trainer.py]: I think it's mainly to have a record as a JSON file associated with each experiment.
Can we maybe find where it fails (convert our params into a dict or similar)?

File: src/transformers/integrations.py (hunk 3 above)
Comment (AyushExel) [reply src/transformers/integrations.py]: Done. please review

File: src/transformers/trainer.py (hunk 6 above)
Comment (AyushExel) [reply src/transformers/trainer.py]: I just added an empty assignment dict in the wandb.config to get by this error

**Comment by AyushExel** [comment]
@borisdayma @morganmcg1 I've tagged you guys in specific places.

File: src/transformers/integrations.py [hunk 7]
```diff
@@ -337,6 +341,61 @@ def run_hp_search_sigopt(trainer, n_trials: int, direction: str, **kwargs) -> Be
...
+    metric = kwargs.pop("metric", None) or "eval/loss"
+
+    sweep_config = trainer.hp_space(None)
+    sweep_config["metric"]["goal"] = direction
+    sweep_config["metric"]["name"] = metric
+    sweep_config["name"] = name
+
+    def _objective():
+        
+        run = wandb.run if wandb.run else wandb.init()
```
Comment (AyushExel) [inline src/transformers/integrations.py:374]: @morganmcg1 @borisdayma 
* Note 1: This uses `WandbCallback` to report metrics. But one `wandb.init` call was necessary to get `wandb.run` object. It won't spin up new run as init is ignored with sweeps.

File: src/transformers/integrations.py [hunk 8]
```diff
@@ -337,6 +341,61 @@ def run_hp_search_sigopt(trainer, n_trials: int, direction: str, **kwargs) -> Be
...
+        trainer.objective = None
+        import pdb
+        trainer.train(resume_from_checkpoint=None, trial=vars(config)['_items'])
+        # If there hasn't been any evaluation during the training loop.
+        if getattr(trainer, "objective", None) is None:
+            metrics = trainer.evaluate()
+            trainer.objective = trainer.compute_objective(metrics)
+            format_metrics = rewrite_logs(metrics)
+            if metric not in format_metrics:
+                raise Exception("Provided metric not found. This might result in expected sweeps charts. Please check the `metric` parameter")
```
Comment (AyushExel) [inline src/transformers/integrations.py]: @morganmcg1 @borisdayma 
* Note 3: Instead of raising an exception, we can also just warn the user. Because they can still select any metric from the UI to see parameter importance and other charts. Only the metric. vs time chart will be broken

File: src/transformers/integrations.py [hunk 9]
```diff
@@ -337,6 +341,61 @@ def run_hp_search_sigopt(trainer, n_trials: int, direction: str, **kwargs) -> Be
...
+    # add WandbCallback if not already added in trainer callbacks
+    if WandbCallback not in trainer.callback_handler.callbacks:
+        trainer.add_callback(WandbCallback)
+    trainer.args.report_to = 'wandb'
+    best_trial = {"run_id": None, "objective": None, "hyperparameters": None}
+    sweep_id = kwargs.pop("sweep_id", None)
+    project = kwargs.pop("project", None)
+    name = kwargs.pop("name", None)
+    entity = kwargs.pop("entity", None)
+    metric = kwargs.pop("metric", None) or "eval/loss"
```
Comment (AyushExel) [inline src/transformers/integrations.py:364]: @morganmcg1 @borisdayma 
* Note 2: user needs to pass `metric` to run sweeps w.r.t to a particular metric. By default, it is `'eval/loss'`

File: src/transformers/trainer.py [hunk 10]
```diff
@@ -1758,6 +1767,8 @@ def hyperparameter_search(
             )
         if backend == HPSearchBackend.SIGOPT and not is_sigopt_available():
             raise RuntimeError("You picked the sigopt backend, but it is not installed. Use `pip install sigopt`.")
+        if backend == HPSearchBackend.WANDB and not is_wandb_available():
+            raise RuntimeError("You picked the sigopt backend, but it is not installed. Use `pip install wandb`.")
```
Comment (morganmcg1) [inline src/transformers/trainer.py]: nit, should be `"You picked the wandb backend..."`

File: src/transformers/trainer.py [hunk 11]
```diff
@@ -1555,6 +1560,10 @@ def _save_checkpoint(self, model, trial, metrics=None):
                 run_id = tune.get_trial_id()
             elif self.hp_search_backend == HPSearchBackend.SIGOPT:
                 run_id = trial.id
+            elif self.hp_search_backend == HPSearchBackend.WANDB:
+                import wandb
```
Comment (morganmcg1) [inline src/transformers/trainer.py:1570]: Do you need to test if wandb is installed, like we do in the [WandbCallback](https://github.com/huggingface/transformers/blob/8b32578119b7b13ccd09772ddcf46de141cd6f27/src/transformers/integrations.py#L468)

Something like:
```
from .integrations import is_wandb_available

...

elif self.hp_search_backend == HPSearchBackend.WANDB:
        has_wandb = is_wandb_available()
        if not has_wandb:
            raise RuntimeError("WandbCallback requires wandb to be installed. Run `pip install wandb`.")
        if has_wandb:
            import wandb
```

File: src/transformers/integrations.py (hunk 8 above)
Comment (morganmcg1) [inline src/transformers/integrations.py]: This could be worded a little clearer:

```
"This might result in expected sweeps"
```

File: src/transformers/trainer_utils.py [hunk 12]
```diff
@@ -211,16 +211,36 @@ def default_hp_space_sigopt(trial):
     ]
 
 
+def default_hp_space_wandb(trial) -> Dict[str, float]:
+    from .integrations import is_wandb_available
+
+    assert is_wandb_available(), "This function needs ray installed: `pip " "install wandb`"
```
Comment (morganmcg1) [inline src/transformers/trainer_utils.py]: nit, should be `"This function needs wandb installed"`

**Review by morganmcg1 — COMMENTED** [review]
just a few high level comments for now, will take another look tomorrow

File: src/transformers/trainer.py (hunk 11 above)
Comment (AyushExel) [reply src/transformers/trainer.py:1570]: It'll be redundant here, this line will be called first https://github.com/AyushExel/transformers/pull/1/files#diff-ed55888e6665791fe92cc8fc0c499da54f4ace6738551cd9a2591881cda076deR1770
so if wandb is not installed, the program won't reach this line

File: src/transformers/trainer_utils.py (hunk 12 above)
Comment (AyushExel) [reply src/transformers/trainer_utils.py]: lol

File: src/transformers/trainer.py (hunk 10 above)
Comment (AyushExel) [reply src/transformers/trainer.py]: lol2

File: src/transformers/integrations.py (hunk 8 above)
Comment (AyushExel) [reply src/transformers/integrations.py]: will do

**Comment by borisdayma** [comment]
Can't see where I'm tagged?

**Comment by AyushExel** [comment]
@borisdayma I don't know why you can't see. I have tagged you for review

Screenshot from 2021-11-22 15-04-56

File: src/transformers/integrations.py (hunk 7 above)
Comment (borisdayma) [reply src/transformers/integrations.py:374]: Is there actually a call to `wandb.init()` at all or will `wandb.run` always return something?

File: src/transformers/integrations.py [hunk 13]
```diff
@@ -337,6 +341,61 @@ def run_hp_search_sigopt(trainer, n_trials: int, direction: str, **kwargs) -> Be
     return best_run
 
 
+def run_hp_search_wandb(trainer, n_trials: int, direction: str, **kwargs) -> BestRun:
+    import wandb
+    # add WandbCallback if not already added in trainer callbacks
+    if WandbCallback not in trainer.callback_handler.callbacks:
+        trainer.add_callback(WandbCallback)
```
Comment (borisdayma) [inline src/transformers/integrations.py]: Pretty good! Does it get added before that when using `report_to=wandb`?

File: src/transformers/integrations.py [hunk 14]
```diff
@@ -337,6 +341,61 @@ def run_hp_search_sigopt(trainer, n_trials: int, direction: str, **kwargs) -> Be
     return best_run
 
 
+def run_hp_search_wandb(trainer, n_trials: int, direction: str, **kwargs) -> BestRun:
+    import wandb
```
Comment (borisdayma) [inline src/transformers/integrations.py:348]: Do we need to check `is_wandb_available()`?

File: src/transformers/integrations.py (hunk 8 above)
Comment (borisdayma) [reply src/transformers/integrations.py]: I'm good with it.
In the warning can you also list provided metric and all metrics available?

File: src/transformers/trainer.py (hunk 10 above)
Comment (borisdayma) [inline src/transformers/trainer.py]: Should it be `wandb` backend?

File: src/transformers/trainer_utils.py (hunk 12 above)
Comment (borisdayma) [inline src/transformers/trainer_utils.py]: Typo, should say `wandb`

**Review by borisdayma — COMMENTED** [review]
Looks good, I left a few small comments.

File: src/transformers/integrations.py (hunk 13 above)
Comment (AyushExel) [reply src/transformers/integrations.py]: Yes. When using `report_to=wandb` it gets added before. I thought It'd make more sense to support both the cases

File: src/transformers/integrations.py (hunk 14 above)
Comment (AyushExel) [reply src/transformers/integrations.py:348]: yep

File: src/transformers/integrations.py (hunk 7 above)
Comment (AyushExel) [reply src/transformers/integrations.py:374]: There is a call as required by sweeps but it comes woth a termlog `call to wandb.init is ignored in sweeps`
